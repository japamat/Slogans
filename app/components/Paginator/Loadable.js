/**
 *
 * Asynchronously loads the component for Paginator
 *
 */

import loadable from 'utils/loadable';

export default loadable(() => import('./index'));
