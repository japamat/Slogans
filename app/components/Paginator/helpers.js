/**
 * Helper functions for Paginator component
 */
import React from 'react';

import PageNumber from '../PageNumber';
import { isMobile } from '../../utils/checkMobile';

const getPageBounds = (total, offset, limit) => {
  const totalPages = Math.floor(total / limit);
  let lower = Math.floor(offset / limit + 1) - 4;
  let upper = Math.floor(offset / limit + 1) + 4;
  if (lower <= 0) {
    lower = 1;
    upper = 9;
  } else if (upper > totalPages) {
    lower = totalPages - 8;
    upper = totalPages;
  }
  return { lower, upper };
};

export const pagesToShow = (total, offset, limit, onClick, theme) => {
  if (isMobile()) {
    return (
      <div>
        <PageNumber
          disabled
          color={theme.bg}
          number={Math.floor(offset / limit + 1)}
          key={`page-number-${Math.floor(offset / limit + 1)}`}
        />
        /
        <PageNumber
          disabled
          color={theme.bg}
          number={Math.floor(total / limit)}
          key="page-number-max-apges"
        />
      </div>
    );
  }
  const pageNumbers = [];
  const { lower, upper } = getPageBounds(total, offset, limit);
  for (let i = lower; i < upper + 1; i += 1) {
    pageNumbers.push(
      <PageNumber
        onClick={() => onClick((i - 1) * limit, limit)}
        color={theme.bg}
        number={i}
        active={Math.floor(offset / limit) + 1 === i}
        key={`page-number-${i / limit}`}
      />,
    );
  }
  return pageNumbers;
};
