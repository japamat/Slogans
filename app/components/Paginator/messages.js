/*
 * Paginator Messages
 *
 * This contains all the text for the Paginator component.
 */

import { defineMessages } from 'react-intl';

export const scope = 'app.components.Paginator';

export default defineMessages({
  header: {
    id: `${scope}.header`,
    defaultMessage: 'This is the Paginator component!',
  },
});
