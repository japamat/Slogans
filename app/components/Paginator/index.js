/**
 *
 * Paginator
 * Takes the total number of results, results offset, limit,
 * onClick function, and theme as props.
 *
 */

import React from 'react';
import PropTypes from 'prop-types';
import { ThemeProvider } from 'styled-components';

import Wrapper from './Wrapper';
import Button from '../Button';
import ArrowIcon from '../ArrowIcon';
import { pagesToShow } from './helpers';

function Paginator(props) {
  const { total, offset, limit, onClick, theme, showPageNums } = props;
  const pageNums = showPageNums
    ? pagesToShow(total, offset, limit, onClick, theme)
    : null;

  return (
    <ThemeProvider theme={theme}>
      <Wrapper>
        <Button
          disabled={offset === 0}
          handleRoute={() => onClick(Math.max(0, offset - limit), limit)}
        >
          <ArrowIcon style={{ transform: 'rotate(180deg)' }} />
        </Button>
        {pageNums}
        <Button
          disabled={offset >= total - limit}
          handleRoute={() => onClick(offset + limit, limit)}
        >
          <ArrowIcon />
        </Button>
      </Wrapper>
    </ThemeProvider>
  );
}

Paginator.propTypes = {
  total: PropTypes.number,
  offset: PropTypes.number,
  limit: PropTypes.number,
  onClick: PropTypes.func,
  theme: PropTypes.object,
  showPageNums: PropTypes.bool,
};

export default Paginator;
