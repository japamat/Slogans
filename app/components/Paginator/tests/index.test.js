/**
 *
 * Tests for Paginator
 *
 * @see https://github.com/react-boilerplate/react-boilerplate/tree/master/docs/testing
 *
 */

import React from 'react';
import { render } from 'react-testing-library';
import { IntlProvider } from 'react-intl';

import Paginator from '../index';
import { DEFAULT_LOCALE } from '../../../i18n';

describe('<Paginator />', () => {
  const clickFn = () => {};
  it('Expect to not log errors in console', () => {
    const spy = jest.spyOn(global.console, 'error');
    render(
      <IntlProvider locale={DEFAULT_LOCALE}>
        <Paginator onCLick={clickFn} theme={{}} />
      </IntlProvider>,
    );
    expect(spy).not.toHaveBeenCalled();
  });

  /**
   * Unskip this test to use it
   *
   * @see {@link https://jestjs.io/docs/en/api#testskipname-fn}
   */
  it('Should render and match the snapshot', () => {
    const {
      container: { firstChild },
    } = render(
      <IntlProvider locale={DEFAULT_LOCALE}>
        <Paginator onCLick={clickFn} theme={{}} />
      </IntlProvider>,
    );
    expect(firstChild).toMatchSnapshot();
  });
});
