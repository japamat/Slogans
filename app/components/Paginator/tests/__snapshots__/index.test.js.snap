// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Paginator /> Should render and match the snapshot 1`] = `
<div
  class="Wrapper-sc-1lznsvd-0 heLOpY"
>
  <div
    class="Wrapper-sc-5dsnpv-0 dFDEkS"
  >
    <button
      class="StyledButton-t20ntd-0 jLhOfj"
    >
      <svg
        height="24"
        style="transform: rotate(180deg);"
        width="24"
      >
        <path
          d="M5.88 4.12L13.76 12l-7.88 7.88L8 22l10-10L8 2z"
        />
        <path
          d="M0 0h24v24H0z"
          fill="none"
        />
      </svg>
    </button>
  </div>
  <div
    class="Wrapper-sc-5dsnpv-0 dFDEkS"
  >
    <button
      class="StyledButton-t20ntd-0 jLhOfj"
    >
      <svg
        height="24"
        width="24"
      >
        <path
          d="M5.88 4.12L13.76 12l-7.88 7.88L8 22l10-10L8 2z"
        />
        <path
          d="M0 0h24v24H0z"
          fill="none"
        />
      </svg>
    </button>
  </div>
</div>
`;
